src/modelfree/training/victim_envs.py

```diff
@@ -94,8 +112,65 @@ def get_curry_obs(self, env_idx=None):
             return self._obs[env_idx]
 
 
+class MultiCurryVecEnv(CurryVecEnv):
+    def __init__(self, venv, policies, agent_idx=0, deterministic=False):
+        """Fixes one of the players in a VecMultiEnv, but alternates between policies
+        used to perform response action.
+        :param venv(VecMultiEnv): the environments.
+        :param policies(iterable of Policy): the policies to use for the agent at agent_idx.
+        :param agent_idx(int): the index of the agent that should be fixed.
+        :param policy_selector(func): A function that takes in
+        :return: a new VecMultiEnv with num_agents decremented. It behaves like env but
+                 with all actions at index agent_idx set to those sampled from one policy
+                 within policies"""
+        super().__init__(venv, policies, agent_idx, deterministic)
+
+        self.policies = policies
+        self.num_envs = venv.num_envs
+        self.state_array = [None]*self.num_envs
+        self.current_policy_idx = [None]*self.num_envs
+        self.current_policies = [None]*self.num_envs
+        self.switch_policy()
+
+    def step_async(self, actions):
+        policy_actions = []
+        for i in range(len(self._obs)):
```

Comment:
I think this is correct but seems weird -- why are we tiling the observations and state arrays and then throwing away all but the first result?

This would be clearer if we just made the batch size always one. More efficient solution would be to evaluate each policy once, on all the observations, then just pay attention to the subset. (I *think* `policy.predict` will require the batch size always be the same, but not sure.)

Reply:
This was required in order to get the policy to predict properly; it would error out and complain about things being the wrong shape if you didn't have an obs x num_envs tensor being passed in, because it's designed for predicting across all envs at once.  (I'm not sure that batch_size is actually the issue here as much as num_envs, but perhaps I'm misunderstanding your comment). 

I went with this approach because it seemed slightly less error-prone than applying both policies to all observations and then applying a mask based on which env is using which policy, but you're right that the masking approach would be more computationally efficient, and I could switch to that if desired.

Reply:
I'd prefer the masking solution. We do care about efficiency here: evaluating the policy is a significant fraction of the cost of rollouts, and most of the training time is spent in collecting rollouts.